File: 572_Subtree_of_Another_Tree.java
/*
https://leetcode.com/problems/subtree-of-another-tree/
Given two non-empty binary trees s and t, check whether tree t has exactly the same structure and node values with a
subtree of s. A subtree of s is a tree consists of a node in s and all of this node's descendants.
The tree s could also be considered as a subtree of itself.

Serialise each tree with special symbol for null. Preceed each node with ",".
Time - O(m*n) because contains() performs naive pattern matching. KMP is O(m + n)
Space - O(m + n)
*/

class Solution {
    StringBuilder sb = new StringBuilder();
    public boolean isSubtree(TreeNode s, TreeNode t) {

        serialize(s);
        String sSerial = sb.toString();
        sb.setLength(0);
        serialize(t);
        String tSerial = sb.toString();

        return sSerial.contains(tSerial);
    }

    private void serialize(TreeNode node) {

        sb.append(",");     // preceed each node with "," to encode s = [1]
                            // as ",1,#,#" and so not be a subtree of t = [21]

        if (node == null) {
            sb.append("#"); // assume "#" cannot be a node.val
            return;
        }

        sb.append(Integer.toString(node.val));
        serialize(node.left);
        serialize(node.right);
    }
}

File: 572_Subtree_of_Another_Tree.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/subtree-of-another-tree/
# Given two non-empty binary trees s and t, check whether tree t has exactly the same structure and node values with a
# subtree of s. A subtree of s is a tree consists of a node in s and all of this node's descendants.
# The tree s could also be considered as a subtree of itself.

# Preorder traversal over s and t to serialize. Nodes are separated by commas (e.g. to distinguish 1 and 0 from 10).
# None nodes are identified with special symbol '#'. Check whether serialization of t is contained in s.
# Time - O(m + n)
# Space - O(m + n)

class Solution(object):
    def isSubtree(self, s, t):
        """
        :type s: TreeNode
        :type t: TreeNode
        :rtype: bool
        """
        def serialize(node):
            if not node:
                serial.append("#")
                return
            serial.append(",")
            serial.append(str(node.val))
            serialize(node.left)
            serialize(node.right)

        serial = []         # list so append each symbol in O(1)
        serialize(s)
        s_serial = "".join(serial)
        serial = []
        serialize(t)
        t_serial = "".join(serial)

        return t_serial in s_serial